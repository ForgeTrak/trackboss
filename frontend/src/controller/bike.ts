import { generateHeaders } from './utils';
import {
    DeleteBikeResponse,
    GetBikeListResponse,
    GetBikeResponse,
    PatchBikeRequest,
    PatchBikeResponse,
    PostNewBikeRequest,
    PostNewBikeResponse,
} from '../../../src/typedefs/bike';

export async function createBike(token: string, bikeData: PostNewBikeRequest): Promise<PostNewBikeResponse> {
    const response = await fetch(`${process.env.REACT_APP_API_URL}/api/bike/new`, {
        method: 'POST',
        mode: 'no-cors',
        headers: generateHeaders(token),
        body: JSON.stringify(bikeData),
    });
    return response.json();
}

export async function getBikeList(token: string, membershipID?: number): Promise<GetBikeListResponse> {
    if (membershipID) {
        const idString = membershipID.toString();
        const response = await fetch(`${process.env.REACT_APP_API_URL}/api/bike/list?membershipID=${idString}`, {
            method: 'GET',
            mode: 'cors',
            headers: generateHeaders(token),
        });
        return response.json();
    }
    // else
    const response = await fetch(`${process.env.REACT_APP_API_URL}/api/bike/list`, {
        method: 'GET',
        mode: 'cors',
        headers: generateHeaders(token),
    });
    return response.json();
}

export async function getBike(token: string, bikeID: number): Promise<GetBikeResponse> {
    const response = await fetch(`${process.env.REACT_APP_API_URL}/api/bike/${bikeID}`, {
        method: 'GET',
        mode: 'no-cors',
        headers: generateHeaders(token),
    });
    return response.json();
}

export async function updateBike(
    token: string,
    bikeID: number,
    bikeData: PatchBikeRequest,
): Promise<PatchBikeResponse> {
    const response = await fetch(`${process.env.REACT_APP_API_URL}/api/bike/${bikeID}`, {
        method: 'PATCH',
        mode: 'no-cors',
        headers: generateHeaders(token),
        body: JSON.stringify(bikeData),
    });
    return response.json();
}

export async function deleteBike(token: string, bikeID: number): Promise<DeleteBikeResponse> {
    const response = await fetch(`${process.env.REACT_APP_API_URL}/api/bike/${bikeID}`, {
        method: 'DELETE',
        mode: 'no-cors',
        headers: generateHeaders(token),
    });
    return response.json();
}
