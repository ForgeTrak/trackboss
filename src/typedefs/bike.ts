import { ErrorResponse } from './errorResponse';

export type Bike = {
    bikeId: number,
    year: string,
    make: string,
    model: string,
    membershipAdmin: string
}

export type DeletedBike = {
    bikeId: number
}

export type PostNewBikeRequest = {
    year?: string,
    make?: string,
    model?: string,
    membershipId: number
}

export type PostNewBikeResponse = Bike | ErrorResponse

export type GetBikeRequest = Record<string, never>

export type GetBikeResponse = Bike | ErrorResponse

export type PatchBikeRequest = {
    year?: string,
    make?: string,
    model?: string,
    membershipId?: number
}

export type PatchBikeResponse = Bike | ErrorResponse

export type DeleteBikeRequest = Record<string, never>

export type DeleteBikeResponse = DeletedBike | ErrorResponse

export type GetBikeListRequest = Record<string, never>

export type GetBikeListResponse = Bike[] | ErrorResponse
