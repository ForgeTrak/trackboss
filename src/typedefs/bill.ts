import { ErrorResponse } from './errorResponse';
import { Job } from './job';

export type Bill = {
    billId: number,
    generatedDate: string,
    year: number,
    amount: number,
    amountWithFee: number,
    pointsEarned: number,
    pointsThreshold: number,
    membershipAdmin: string,
    membershipAdminEmail: string,
    phone: string,
    membershipId: number,
    firstName: string,
    lastName: string,
    membershipType: string,
    emailedBill?: string, // undefined if not emailed, date otherwise
    curYearPaid: boolean,
    curYearIns: boolean,
    dueDate: string,
    paymentMethod?: string,
    squareLink?: string,
    squareOrderId?: string,
    detail?: string,
    contactedAndRenewing?: boolean,
    memberActive: boolean,
}

export type WorkPointThreshold = {
    year: number,
    threshold: number
}

export type GetWorkPointThresholdRequest = Record<string, never>

export type GetWorkPointThresholdResponse = WorkPointThreshold | ErrorResponse

export type GetBillListRequest = Record<string, never>

// Used only for the backend: API layer -> DB layer communication
export type GetBillListRequestFilters = {
    membershipId?: number,
    year?: number,
    paymentStatus?: string,
    membershipStatus?: string,
}

export type GetBillListResponse = Bill[] | ErrorResponse

export type PostCalculateBillsRequest = Record<string, never>

export type PostCalculateBillsResponse = Bill[] | ErrorResponse

// Used only for the backend: API layer -> DB layer communication
export type GenerateSingleBillRequest = {
    amount: number,
    amountWithFee: number,
    membershipId: number,
    pointsEarned: number,
    pointsThreshold: number,
    workDetail?: Job[],
    billingYear: number,
}

export type GetMembershipBillListRequest = Record<string, never>

export type GetMembershipBillListResponse = Bill[] | ErrorResponse

export type PostPayBillRequest = Record<string, never>

export type PostPayBillResponse = Record<string, never> | ErrorResponse
