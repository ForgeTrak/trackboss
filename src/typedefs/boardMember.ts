import { ErrorResponse } from './errorResponse';

export type BoardMember = {
    boardId: number,
    title: string,
    year: string,
    memberId: number
}

export type GetBoardMemberRequest = Record<string, never>;

export type GetBoardMemberResponse = BoardMember | ErrorResponse;

export type PostNewBoardMemberRequest = {
    boardMemberTitleId: number,
    year: string,
    memberId: number
}

export type PostNewBoardMemberResponse = BoardMember | ErrorResponse;

export type PatchBoardMemberRequest = {
    boardMemberTitleId?: number,
    year?: string,
    memberId?: number
};

export type PatchBoardMemberResponse = BoardMember | ErrorResponse;

export type GetBoardMemberListRequest = Record<string, never>;

export type GetBoardMemberListResponse = BoardMember[] | ErrorResponse;

export type DeleteBoardMemberRequest = Record<string, never>;

export type DeleteBoardMemberResponse = {
    boardMemberId: number,
} | ErrorResponse;
