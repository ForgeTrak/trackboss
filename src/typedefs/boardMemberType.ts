import { ErrorResponse } from './errorResponse';

export type boardMemberType = {
    boardTypeId: number,
    title: string,
}

export type GetBoardMemberTypeRequest = {};

export type GetBoardMemberTypeResponse = boardMemberType | ErrorResponse;

export type PostNewBoardMemberTypeRequest = {
    title: string
}

export type PostNewBoardMemberTypeResponse = boardMemberType | ErrorResponse;

export type PatchBoardMemberTypeRequest = {
    boardTypeId: number,
    title: string
};

export type PatchBoardMemberTypeResponse = boardMemberType | ErrorResponse;

export type GetBoardMemberTypeListRequest = {};

export type GetBoardMemberTypeListResponse = boardMemberType[] | ErrorResponse;
