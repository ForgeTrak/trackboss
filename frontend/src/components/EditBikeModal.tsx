import React, { useEffect, useState } from 'react';
import {
    Modal,
    ModalOverlay,
    ModalContent,
    ModalCloseButton,
    Button,
    Divider,
    Heading,
    ModalFooter,
    Text,
    VStack,
    Input,

} from '@chakra-ui/react';
import { Bike } from '../../../src/typedefs/bike';

interface modalProps {
  isOpen: boolean,
  onClose: () => void,
  bikeToEdit: Bike,
  // eslint-disable-next-line no-unused-vars
  editBike: (bikeToEdit: Bike, year: string, make: string, model: string) => void,
}

export default function EditBikesModal(props: modalProps) {
    const [bike, setBike] = useState<Bike>();

    const [bikeYear, setBikeYear] = useState<string>('');
    const [bikeMake, setBikeMake] = useState<string>('');
    const [bikeModel, setBikeModel] = useState<string>('');

    const handleEditedBikeYear = (event: { target: { value: any; }; }) => setBikeYear(event.target.value);
    const handleEditedBikeMake = (event: { target: { value: any; }; }) => setBikeMake(event.target.value);
    const handleEditedBikeModel = (event: { target: { value: any; }; }) => setBikeModel(event.target.value);

    useEffect(() => {
        async function setState() {
            setBike(props.bikeToEdit);
        }
        setState();
    }, [props.bikeToEdit]);

    return (
        <Modal isCentered size="xl" isOpen={props.isOpen} onClose={props.onClose}>
            <ModalOverlay />
            <ModalContent>
                <Heading pl={2} pr={2} textAlign="center">
                    Edit this bike
                </Heading>
                <Text fontSize="2xl" textAlign="center">
                    Current Bike Info:
                    {' '}
                    {`${bike?.year}, ${bike?.make} ${bike?.model}`}
                </Text>
                <Divider mb={5} />
                <ModalCloseButton />
                <VStack
                    align="left"
                    mr={5}
                    ml={5}
                >
                    <Input
                        variant="outline"
                        placeholder="Bike year"
                        value={bikeYear}
                        onChange={handleEditedBikeYear}
                        size="md"
                    />
                    <Input
                        variant="outline"
                        placeholder="Bike Make"
                        value={bikeMake}
                        onChange={handleEditedBikeMake}
                        size="md"
                    />
                    <Input
                        variant="outline"
                        placeholder="Bike Model"
                        value={bikeModel}
                        onChange={handleEditedBikeModel}
                        size="md"
                    />
                </VStack>
                <ModalFooter>
                    <Button
                        variant="ghost"
                        mr={3}
                        size="lg"
                        onClick={
                            () => {
                                props.onClose();
                            }
                        }
                    >
                        Close
                    </Button>
                    <Button
                        color="green"
                        variant="ghost"
                        mr={3}
                        size="lg"
                        onClick={
                            () => {
                                props.editBike(props.bikeToEdit, bikeYear, bikeMake, bikeModel);
                                props.onClose();
                            }
                        }
                    >
                        Save
                    </Button>
                </ModalFooter>
            </ModalContent>
        </Modal>
    );
}
