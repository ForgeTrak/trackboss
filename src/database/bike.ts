import _ from 'lodash';
import { OkPacket, RowDataPacket } from 'mysql2';

import logger from '../logger';
import pool from './pool';
import { Bike, PostNewBikeRequest } from '../typedefs/bike';

export const INSERT_BIKE_SQL = 'INSERT INTO member_bikes (year, make, model, membership_id) VALUES (?, ?, ?, ?)';
export const GET_BIKE_SQL = 'SELECT bike_id, year, make, model, membership_admin FROM v_bike WHERE bike_id = ?';

export async function insertBike(req: PostNewBikeRequest): Promise<number> {
    //const sql = 'INSERT INTO member_bikes (year, make, model, membership_id) VALUES (?, ?, ?, ?)';
    const values = [req.year, req.make, req.model, req.membershipId];

    let result;
    try {
        [result] = await pool.query<OkPacket>(INSERT_BIKE_SQL, values);
    } catch (e) {
        logger.error(`DB error inserting bike: ${e}`);
        throw new Error('internal server error');
    }

    return result.insertId;
}

export async function getBike(id: number): Promise<Bike> {
    //const sql = 'SELECT bike_id, year, make, model, membership_admin FROM v_bike WHERE bike_id = ?';
    const values = [id];

    let results;
    try {
        [results] = await pool.query<RowDataPacket[]>(GET_BIKE_SQL, values);
    } catch (e) {
        logger.error(`DB error getting bike: ${e}`);
        throw new Error('internal server error');
    }

    if (_.isEmpty(results)) {
        throw new Error('not found');
    }

    return {
        bikeId: results[0].bike_id,
        year: results[0].year,
        make: results[0].make,
        model: results[0].model,
        membershipAdmin: results[0].membership_admin,
    };
}
