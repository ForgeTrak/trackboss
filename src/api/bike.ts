import { Request, Response, Router } from 'express';
import { deleteBike, getBike, getBikeList, insertBike, patchBike } from '../database/bike';
import {
    Bike,
    DeleteBikeResponse,
    GetBikeListResponse,
    GetBikeResponse,
    PatchBikeResponse,
    PostNewBikeResponse,
} from '../typedefs/bike';
import { checkHeader, verify } from '../util/auth';

const bike = Router();

bike.post('/new', async (req: Request, res: Response) => {
    const { authorization } = req.headers;
    let response: PostNewBikeResponse;
    const headerCheck = checkHeader(authorization);
    if (!headerCheck.valid) {
        res.status(401);
        response = { reason: headerCheck.reason };
    } else {
        try {
            await verify(headerCheck.token, 'Membership Admin');
            const insertId = await insertBike(req.body);
            response = await getBike(insertId);
            res.status(201);
        } catch (e: any) {
            if (e.message === 'user input error') {
                res.status(400);
                response = { reason: 'bad request' };
            } else if (e.message === 'Authorization Failed') {
                res.status(401);
                response = { reason: 'not authorized' };
            } else if (e.message === 'Forbidden') {
                res.status(403);
                response = { reason: 'forbidden' };
            } else {
                res.status(500);
                response = { reason: 'internal server error' };
            }
        }
    }
    res.send(response);
});

bike.get('/list', async (req: Request, res: Response) => {
    const { authorization } = req.headers;
    let response: GetBikeListResponse;
    const headerCheck = checkHeader(authorization);
    if (!headerCheck.valid) {
        res.status(401);
        response = { reason: headerCheck.reason };
    } else {
        try {
            await verify(headerCheck.token);
            const id = req.query.membershipID;
            let filterMembership: number | undefined;
            if (typeof id === 'undefined') {
                filterMembership = undefined;
            } else {
                filterMembership = Number(id);
                if (Number.isNaN(filterMembership)) {
                    throw new Error('user input error');
                }
            }
            const bikeList: Bike[] = await getBikeList(filterMembership);
            res.status(200);
            response = bikeList;
        } catch (e: any) {
            if (e.message === 'user input error') {
                res.status(400);
                response = { reason: 'bad request' };
            } else if (e.message === 'Authorization Failed') {
                res.status(401);
                response = { reason: 'not authorized' };
            } else {
                res.status(500);
                response = { reason: 'internal server error' };
            }
        }
    }
    res.send(response);
});

bike.get('/:bikeID', async (req: Request, res: Response) => {
    const { authorization } = req.headers;
    let response: GetBikeResponse;
    const headerCheck = checkHeader(authorization);
    if (!headerCheck.valid) {
        res.status(401);
        response = { reason: headerCheck.reason };
    } else {
        try {
            await verify(headerCheck.token);
            const { bikeID } = req.params;
            response = await getBike(Number(bikeID));
            res.status(200);
        } catch (e: any) {
            if (e.message === 'not found') {
                res.status(404);
                response = { reason: 'not found' };
            } else if (e.message === 'Authorization Failed') {
                res.status(401);
                response = { reason: 'not authorized' };
            } else {
                res.status(500);
                response = { reason: 'internal server error' };
            }
        }
    }
    res.send(response);
});

bike.patch('/:bikeID', async (req: Request, res: Response) => {
    const { authorization } = req.headers;
    let response: PatchBikeResponse;
    const headerCheck = checkHeader(authorization);
    if (!headerCheck.valid) {
        res.status(401);
        response = { reason: headerCheck.reason };
    } else {
        try {
            const { bikeID } = req.params;
            const bikeIdNum = Number(bikeID);
            if (Number.isNaN(bikeIdNum)) {
                throw new Error('not found');
            }
            await verify(headerCheck.token, 'Membership Admin');
            await patchBike(bikeIdNum, req.body);
            response = await getBike(bikeIdNum);
            res.status(200);
        } catch (e: any) {
            if (e.message === 'user input error') {
                res.status(400);
                response = { reason: 'bad request' };
            } else if (e.message === 'not found') {
                res.status(404);
                response = { reason: 'not found' };
            } else if (e.message === 'Authorization Failed') {
                res.status(401);
                response = { reason: 'not authorized' };
            } else if (e.message === 'Forbidden') {
                res.status(403);
                response = { reason: 'forbidden' };
            } else {
                res.status(500);
                response = { reason: 'internal server error' };
            }
        }
    }
    res.send(response);
});

bike.delete('/:bikeID', async (req: Request, res: Response) => {
    const { authorization } = req.headers;
    let response: DeleteBikeResponse;
    const headerCheck = checkHeader(authorization);
    if (!headerCheck.valid) {
        res.status(401);
        response = { reason: headerCheck.reason };
    } else {
        try {
            const { bikeID } = req.params;
            const bikeIdNum = Number(bikeID);
            if (Number.isNaN(bikeIdNum)) {
                throw new Error('not found');
            }
            await verify(headerCheck.token, 'Membership Admin');
            await deleteBike(bikeIdNum);
            response = { bikeId: bikeIdNum };
            res.status(200);
        } catch (e: any) {
            if (e.message === 'not found') {
                res.status(404);
                response = { reason: 'not found' };
            } else if (e.message === 'Authorization Failed') {
                res.status(401);
                response = { reason: 'not authorized' };
            } else if (e.message === 'Forbidden') {
                res.status(403);
                response = { reason: 'forbidden' };
            } else {
                res.status(500);
                response = { reason: 'internal server error' };
            }
        }
    }
    res.send(response);
});

export default bike;
