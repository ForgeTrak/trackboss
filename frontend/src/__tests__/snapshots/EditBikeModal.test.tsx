import React from 'react';
import { BrowserRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import { Bike } from '../../../../src/typedefs/bike';
import EditBikeModal from '../../components/EditBikeModal';

function edit() {
    // void function
}

function onClose() {
    // void function
}

const bike: Bike = {
    bikeId: 1,
    membershipAdmin: 'me',
    year: '2000',
    make: 'Honda',
    model: '250',
};

// Necessary to mock the chakra-ui modal portal, so that the renderer can target the modal instead of the page
const divWithChildrenMock = (children: any, identifier: any) => <div data-testId={identifier}>{children}</div>;
jest.mock('@chakra-ui/react', () => (
    {
        ...jest.requireActual('@chakra-ui/react'),
        PortalManager: jest.fn(({ children }) => divWithChildrenMock(children, 'portal')),
    }
));

describe('edit bike modal', () => {
    it('renders correctly', () => {
        const header = renderer.create(
            <BrowserRouter>
                <EditBikeModal bikeToEdit={bike} isOpen onClose={onClose} editBike={edit} />
            </BrowserRouter>,
        ).toJSON();
        expect(header).toMatchSnapshot();
    });
});
