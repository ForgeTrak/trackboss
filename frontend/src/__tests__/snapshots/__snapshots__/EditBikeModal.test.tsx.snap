// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`edit bike modal renders correctly 1`] = `<span />`;
